File: music-core/src/main/java/mobile/substance/sdk/music/core/objects/Album.java
/*
 * Copyright 2016 Substance Mobile
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package mobile.substance.sdk.music.core.objects;

import android.graphics.Bitmap;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.v4.media.MediaMetadataCompat;
import android.util.Log;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;

import mobile.substance.sdk.music.core.MusicCoreOptions;

public class Album extends MediaObject {

    @Override
    protected Uri getBaseUri() {
        return MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Title
    ///////////////////////////////////////////////////////////////////////////

    public String getAlbumName() {
        return data.getString(MediaMetadataCompat.METADATA_KEY_TITLE);
    }

    public void setAlbumName(String albumName) {
        putString(MediaMetadataCompat.METADATA_KEY_TITLE, albumName);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Album Art
    ///////////////////////////////////////////////////////////////////////////

    private String albumArtworkPath;
    private Uri albumArtworkUri;

    public void setAlbumArtworkPath(String artworkPath) {
        this.albumArtworkPath = artworkPath;
        this.albumArtworkUri = Uri.parse("file://" + artworkPath);
        if (artworkPath != null) {
            setAnimated(false);
        } else {
            setAnimated(true);
        }
    }

    public String getAlbumArtworkPath() {
        return albumArtworkPath == null ? "" : albumArtworkPath;
    }

    public Uri getAlbumArtworkUri() {
        return albumArtworkUri;
    }

    public interface ArtRequest {
        void respond(Bitmap albumArt);
    }

    public void requestArt(final ArtRequest request) {
        Glide.with(getContext()).load(getAlbumArtworkPath())
                .asBitmap()
                .placeholder(MusicCoreOptions.getDefaultArt())
                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                .animate(android.R.anim.fade_in)
                .centerCrop()
                .into(new SimpleTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
                        request.respond(resource);
                    }
                });
        Log.d("requestArt(ArtRequest)", getAlbumArtworkPath());
    }

    public void requestArt(ImageView imageView) {
        Glide.with(getContext()).load(getAlbumArtworkPath())
                .placeholder(MusicCoreOptions.getDefaultArt())
                .diskCacheStrategy(DiskCacheStrategy.SOURCE)
                .crossFade()
                .centerCrop()
                .into(imageView);
        Log.d("requestArt(ImageView)", getAlbumArtworkPath());
    }

    ///////////////////////////////////////////////////////////////////////////
    // Artist
    ///////////////////////////////////////////////////////////////////////////

    public String getAlbumArtistName() {
        return data.getString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST);
    }

    public void setAlbumArtistName(String albumArtistName) {
        putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST, albumArtistName);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Genre
    ///////////////////////////////////////////////////////////////////////////

    public String getAlbumGenreName() {
        return data.getString(MediaMetadataCompat.METADATA_KEY_GENRE);
    }

    public void setAlbumGenreName(String albumGenreName) {
        putString(MediaMetadataCompat.METADATA_KEY_GENRE, albumGenreName);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Year
    ///////////////////////////////////////////////////////////////////////////

    public long getAlbumYear() {
        return data.getLong(MediaMetadataCompat.METADATA_KEY_YEAR);
    }

    public void setAlbumYear(long albumYear) {
        putLong(MediaMetadataCompat.METADATA_KEY_YEAR, albumYear);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Number Of Songs
    ///////////////////////////////////////////////////////////////////////////

    public long getAlbumNumberOfSongs() {
        return data.getLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS);
    }

    public void setAlbumNumberOfSongs(long albumNumberOfSongs) {
        putLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS, albumNumberOfSongs);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Context behavior
    ///////////////////////////////////////////////////////////////////////////

    @Override
    protected boolean isContextRequired() {
        return true;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Color Holding
    ///////////////////////////////////////////////////////////////////////////

    public void setColors(Object colors) {
        putData("album_color", colors);
    }

    public Object getColors() {
        return getData("album_color");
    }

    ///////////////////////////////////////////////////////////////////////////
    // Builder
    ///////////////////////////////////////////////////////////////////////////

    public static class Builder {
        private Album album;

        public Builder() {
            this(new Album());
        }

        public Builder(Album copy) {
            this.album = copy;
        }

        public Builder setName(String name) {
            this.album.setAlbumName(name);
            return this;
        }

        public Builder setArtistName(String artistName) {
            this.album.setAlbumArtistName(artistName);
            return this;
        }

        public Builder setGenreName(String genreName) {
            this.album.setAlbumGenreName(genreName);
            return this;
        }

        public Builder setYear(long year) {
            this.album.setAlbumYear(year);
            return this;
        }

        public Builder setAlbumId(long id) {
            this.album.setID(id);
            return this;
        }

        public Builder setNumberOfSongs(int numberOfSongs) {
            this.album.setAlbumNumberOfSongs(numberOfSongs);
            return this;
        }

        public Builder setArtworkPath(String path) {
            this.album.setAlbumArtworkPath(path);
            return this;
        }

        public Album build() {
            return album;
        }
    }
}

File: music-core/src/main/kotlin/mobile/substance/sdk/music/core/objects/Album.kt
/*
 * Copyright 2016 Substance Mobile
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package mobile.substance.sdk.music.core.objects

import android.graphics.Bitmap
import android.net.Uri
import android.provider.MediaStore
import android.support.v4.media.MediaMetadataCompat
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.SimpleTarget
import mobile.substance.sdk.music.core.MusicCoreOptions

class Album : MediaObject() {

    ///////////////////////////////////////////////////////////////////////////
    // Overrides
    ///////////////////////////////////////////////////////////////////////////

    override val baseUri: Uri?
        get() = MediaStore.Audio.Albums.EXTERNAL_CONTENT_URI

    override val isContextRequired: Boolean
        get() = true

    ///////////////////////////////////////////////////////////////////////////
    // Title
    ///////////////////////////////////////////////////////////////////////////

    var albumName: String?
        get() = metadata?.getString(MediaMetadataCompat.METADATA_KEY_TITLE)
        set(value) {
            if(value != null) putString(MediaMetadataCompat.METADATA_KEY_TITLE, value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Album Artwork
    ///////////////////////////////////////////////////////////////////////////

    var albumArtworkPath: String?
        get() {
            val path = metadata?.getString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI)
            if(path != null) return path else return ""
        }
        set(value) {
            if (value != null) {
                putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ART_URI, value)
                albumArtworkUri = Uri.parse("file://" + value)
                isAnimated = false
            } else {
                isAnimated = true
            }
        }

    var albumArtworkUri: Uri? = null
        private set


    interface ArtRequest {
        fun respond(albumArt: Bitmap)
    }

    fun requestArt(request: ArtRequest) {
        Glide.with(getContext()).load(albumArtworkPath).asBitmap().placeholder(MusicCoreOptions.defaultArt).diskCacheStrategy(DiskCacheStrategy.SOURCE).animate(android.R.anim.fade_in).centerCrop().into(object : SimpleTarget<Bitmap>() {
            override fun onResourceReady(resource: Bitmap, glideAnimation: GlideAnimation<in Bitmap>) {
                request.respond(resource)
            }
        })
    }

    fun requestArt(imageView: ImageView) {
        Glide.with(getContext()).load(albumArtworkPath).placeholder(MusicCoreOptions.defaultArt).diskCacheStrategy(DiskCacheStrategy.SOURCE).crossFade().centerCrop().into(imageView)
    }

    ///////////////////////////////////////////////////////////////////////////
    // Artist
    ///////////////////////////////////////////////////////////////////////////

    var albumArtistName: String?
        get() = metadata?.getString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST)
        set(value) {
            if(value != null) putString(MediaMetadataCompat.METADATA_KEY_ALBUM_ARTIST, value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Genre
    ///////////////////////////////////////////////////////////////////////////

    var albumGenreName: String?
        get() = metadata?.getString(MediaMetadataCompat.METADATA_KEY_GENRE)
        set(value) {
            if(value != null) putString(MediaMetadataCompat.METADATA_KEY_GENRE, value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Year
    ///////////////////////////////////////////////////////////////////////////

    var albumYear: Long?
        get() = metadata?.getLong(MediaMetadataCompat.METADATA_KEY_YEAR)
        set(value) {
            if(value != null) putLong(MediaMetadataCompat.METADATA_KEY_YEAR, value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Number Of Songs
    ///////////////////////////////////////////////////////////////////////////

    var albumNumberOfSongs: Long?
        get() = metadata?.getLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS)
        set(value) {
            if(value != null) putLong(MediaMetadataCompat.METADATA_KEY_NUM_TRACKS, value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Color Holding
    ///////////////////////////////////////////////////////////////////////////

    var colors: Any?
        get() = getData("album_colors")!!
        set(value) {
            if(value != null) putData("album_colors", value)
        }

    ///////////////////////////////////////////////////////////////////////////
    // Builder
    ///////////////////////////////////////////////////////////////////////////

    class Builder @JvmOverloads constructor(private val album: Album = Album()) {

        fun setAlbumId(id: Long): Builder {
            this.album.id = id
            return this
        }

        fun setName(name: String): Builder {
            this.album.albumName = name
            return this
        }

        fun setArtistName(artistName: String): Builder {
            this.album.albumArtistName = artistName
            return this
        }

        fun setGenreName(genreName: String): Builder {
            this.album.albumGenreName = genreName
            return this
        }

        fun setYear(year: Long): Builder {
            this.album.albumYear = year
            return this
        }

        fun setNumberOfSongs(numberOfSongs: Long): Builder {
            this.album.albumNumberOfSongs = numberOfSongs
            return this
        }

        fun setArtworkPath(path: String?): Builder {
            this.album.albumArtworkPath = path
            return this
        }

        fun build(): Album {
            return album
        }
    }
}


